### ISSUE TITLE
Question on the flow normalization

### ISSUE THREAD

**Issue by ghost** [ucbdrive/hd3#11]
Hello,
Could I ask why you normalize the flow by
vgrid[:,0,:,:] = 2.0*vgrid[:,0,:,:].clone() / max(W-1,1)-1.0 in the warping function?

Due to minus value in flow, the vgrid could also have minus and then the result of that line could be less then -1. Isnt that out of purposed vgrid range [-1, 1]?

Can we assume that vgrid == sum(grid, flow) always stay larger than 0?

**Comment by yzcjtr** [comment]
Hi, the normalization operation transforms sampled coordinates from [0, W-1]*[0, H-1] to [-1, 1] * [-1,1]. The normalized flow certainly contains values less than -1, in which case the original coordinate should be less than 0 and the sampler would use border values as padding.

**Comment by ghost** [comment]
Thanks for replying quickly.

If the main purpose is normalizing coord into [-1, 1], will it be better to
normalize coord and flow separately into the range first then, add them
together?
currently with the coord focused normalizing , it seems large negative flows like -10,
-30 all become -1.x and lose their intensity by changed with padding value.

Is there any intention or benefit that i missed why "add an unnormalized [-x, x] flow into [0,
h] coord then normalize together" instead of "normalize separately and then
add together"?

Thanks

**Comment by yzcjtr** [comment]
The purpose of the normalization is to make the sampling coordinates compatible with PyTorch's bilinear sampler API. I would suggest you figure out the mechanism of differentiable warping guided by optical flow first.

**Comment by ghost** [comment]
Thanks! I will read that.
Could you help with one more guide for starter: what is an expected range of the flow when it given as input to the warping function?

**Comment by ghost** [comment]
Thanks! I will read that.
Could you help with one more guide for starter: what is an expected range of the flow when it given as input to the warping function?

**Comment by yzcjtr** [comment]
You can find this at https://pytorch.org/docs/1.0.0/nn.html?highlight=grid_sample#torch.nn.functional.grid_sample.
